src/Evolution/Systems/ScalarWave/UpwindPenaltyCorrection.cpp

```diff
@@ -0,0 +1,168 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/ScalarWave/UpwindPenaltyCorrection.hpp"
+
+#include <algorithm>
+#include <array>
+
+#include "DataStructures/DataBox/PrefixHelpers.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"  // IWYU pragma: keep
+#include "Evolution/Systems/ScalarWave/System.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"   // IWYU pragma: keep
+#include "Utilities/TMPL.hpp"  // IWYU pragma: keep
+
+/// \cond
+namespace ScalarWave {
+template <size_t Dim>
+void UpwindPenaltyCorrection<Dim>::package_data(
+    const gsl::not_null<Scalar<DataVector>*> packaged_v_psi,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        packaged_v_zero,
+    const gsl::not_null<Scalar<DataVector>*> packaged_v_plus,
+    const gsl::not_null<Scalar<DataVector>*> packaged_v_minus,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        packaged_n_times_v_plus,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        packaged_n_times_v_minus,
+    const gsl::not_null<Scalar<DataVector>*> packaged_gamma2_v_psi,
+    const gsl::not_null<tnsr::a<DataVector, 3, Frame::Inertial>*>
+        packaged_char_speeds,
+
+    const Scalar<DataVector>& v_psi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& v_zero,
+    const Scalar<DataVector>& v_plus, const Scalar<DataVector>& v_minus,
+    const std::array<DataVector, 4>& char_speeds,
+    const Scalar<DataVector>& constraint_gamma2,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& interface_unit_normal)
+    const noexcept {
+  // Computes the contribution to the numerical flux from one side of the
+  // interface.
+  //
+  // Note: when PenaltyFlux::operator() is called, an Element passes in its own
+  // packaged data to fill the interior fields, and its neighbor's packaged data
+  // to fill the exterior fields. This introduces a sign flip for each normal
+  // used in computing the exterior fields.
+  get(*packaged_v_psi) = char_speeds[0] * get(v_psi);
```

Comment:
As in the GH penalty flux, would be nice to name these more accurately. E.g. weighted_v_psi or char_speed_weighted_v_psi or something, since otherwise a reader might think this variable just holds v_psi, and not v_psi. Same comment for the other variables

But again, I don't want to delay the PR, so please consider this just a suggestion, if you can do it without causing too much delay

Reply:
+1